src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs

```diff
@@ -678,6 +717,84 @@ public void TestMergeChanges_Overlapping_OldBeforeNew()
+            var change1 = original.WithChanges(new TextChange(new TextSpan(1, 3), "aa"));
+            var change2 = change1.WithChanges(new TextChange(new TextSpan(1, 0), "bb"));
+
+            var changes = change2.GetTextChanges(original);
+            Assert.Equal("0aa4", change1.ToString());
+            Assert.Equal("0bbaa4", change2.ToString());
+            Assert.Equal(new[] { new TextChange(new TextSpan(1, 3), "bbaa") }, changes);
+        }
+
+        [Fact]
+        public void TestMergeChanges_SameStart_AndNewHasDeletion_SmallerThanOldInsertion()
+        {
+            var original = SourceText.From("01234");
+            var change1 = original.WithChanges(new TextChange(new TextSpan(1, 0), "aa"));
+            var change2 = change1.WithChanges(new TextChange(new TextSpan(1, 1), "bb"));
+
+            var changes = change2.GetTextChanges(original);
+            Assert.Equal("0aa1234", change1.ToString());
+            Assert.Equal("0bba1234", change2.ToString());
+            Assert.Equal(new[] { new TextChange(new TextSpan(1, 0), "bba") }, changes);
+        }
+
+        [Fact]
+        public void TestMergeChanges_SameStart_AndNewHasDeletion_EqualToOldInsertion()
+        {
+            var original = SourceText.From("01234");
+            var change1 = original.WithChanges(new TextChange(new TextSpan(1, 0), "aa"));
+            var change2 = change1.WithChanges(new TextChange(new TextSpan(1, 2), "bb"));
+
+            var changes = change2.GetTextChanges(original);
+            Assert.Equal("0aa1234", change1.ToString());
+            Assert.Equal("0bb1234", change2.ToString());
+            Assert.Equal(new[] { new TextChange(new TextSpan(1, 0), "bb") }, changes);
+        }
+
+        [Fact]
+        public void TestMergeChanges_SameStart_AndNewHasDeletion_LargerThanOldInsertion()
+        {
+            var original = SourceText.From("01234");
+            var change1 = original.WithChanges(new TextChange(new TextSpan(1, 0), "aa"));
+            var change2 = change1.WithChanges(new TextChange(new TextSpan(1, 3), "bb"));
+
+            var changes = change2.GetTextChanges(original);
+            Assert.Equal("0aa1234", change1.ToString());
+            Assert.Equal("0bb234", change2.ToString());
+            Assert.Equal(new[] { new TextChange(new TextSpan(1, 1), "bb") }, changes);
+        }
+
+        [Fact]
+        public void TestMergeChanges_SameStart_AndBothHaveDeletion_NewDeletionSmallerThanOld()
+        {
+            var original = SourceText.From("01234");
+            var change1 = original.WithChanges(new TextChange(new TextSpan(1, 3), "aa"));
+            var change2 = change1.WithChanges(new TextChange(new TextSpan(1, 1), "bb"));
+
+            var changes = change2.GetTextChanges(original);
+            Assert.Equal("0aa4", change1.ToString());
+            Assert.Equal("0bba4", change2.ToString());
+            Assert.Equal(new[] { new TextChange(new TextSpan(1, 3), "bba") }, changes);
+        }
```

Comment:
>} [](start = 8, length = 1)

Do we also need to handle the case where the new deletion is larger than the old? It looks like the following fails with or without this PR:
```C#
var original = SourceText.From("01234");
var change1 = original.WithChanges(new TextChange(new TextSpan(1, 3), "aa"));
var change2 = change1.WithChanges(new TextChange(new TextSpan(1, 3), "bb"));

var changes = change2.GetTextChanges(original);
Assert.Equal("0aa4", change1.ToString());
Assert.Equal("0bb", change2.ToString());
```
We can log a bug for this case and handle it separately.

Reply:
Thanks! I'll take a look in a few moments -- if it's easy to fix then maybe just fix it here, if hard (needs a lot of review) then I can do another PR.

Reply:
Seems similar to what I am already doing in `newChange.Span.Start > oldChange.Span.Start + oldDelta`, but that means it's complicated enough for it's own review.

So I would prefer to do as a separate bug/PR just to make review easier (also maybe there is some reuse that can be done between those two cases).